core/services/gateway/connectionmanager.go

```diff
@@ -254,3 +263,34 @@ func (m *donConnectionManager) readLoop(nodeAddress string, nodeState *nodeState
 		}
 	}
 }
+
+func (m *donConnectionManager) heartbeatLoop(intervalSec uint32) {
+	ctx, _ := utils.StopChan(m.shutdownCh).NewCtx()
+	defer m.closeWait.Done()
+
+	if intervalSec == 0 {
+		m.lggr.Error("heartbeat interval is 0, heartbeat disabled")
+		return
+	}
+	m.lggr.Info("starting heartbeat loop")
+
+	ticker := time.NewTicker(time.Duration(intervalSec) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.shutdownCh:
+			return
+		case <-ticker.C:
+			errorCount := 0
+			for nodeAddress, nodeState := range m.nodes {
+				err := nodeState.conn.Write(ctx, websocket.PingMessage, []byte{})
+				if err != nil {
+					m.lggr.Debugw("unable to send heartbeat to node", "nodeAddress", nodeAddress, "err", err)
+					errorCount++
+				}
+			}
+			m.lggr.Infow("sent heartbeat to nodes", "donID", m.donConfig.DonId, "errCount", errorCount)
```

Comment:
Maybe excessive logging too.. 
A better solution is to use prometheus counters for successful sent, errors count and received pings. Instead of logging.

Reply:
This will happen only every 30 seconds. I'll move to a prom counter when that's ready.